Zend/zend_compile.c

```diff
@@ -4637,6 +4637,9 @@ void zend_begin_method_decl(zend_op_array *op_array, zend_string *name, zend_boo
 	} else if (!has_body) {
 		zend_error_noreturn(E_COMPILE_ERROR, "Non-abstract method %s::%s() must contain body",
 			ZSTR_VAL(ce->name), ZSTR_VAL(name));
+	} else if (in_trait) {
+		/* force abstract in traits to prevent the_trait::static_method() from being allowed to be called */
+		op_array->fn_flags |= ZEND_ACC_ABSTRACT | ZEND_ACC_CHANGED;
```

Comment:
Do you have a better idea how to make methods non-callable and which integrates nicely into the existing source?